### PR TITLE
feat(product): introduce multiple loyalty prices

### PR DESCRIPTION
**Pull request by IvanMaidurov**

### CODE REVIEW

File: product/domain/productBasics.go [hunk 1]
```diff
@@ -90,6 +90,7 @@ type (
 		SaleableTo      time.Time
 		ActivePrice     PriceInfo
 		AvailablePrices []PriceInfo
+		LoyaltyPrice    LoyaltyPriceInfo
```
Comment (tessig) [inline product/domain/productBasics.go:93]: Why do we need this addition? Is it the active loyalty price? Would call it then like this.

File: product/interfaces/graphql/product/dto/productdto_activevariantproduct.go [hunk 2]
```diff
@@ -89,6 +89,17 @@ func (avp ActiveVariantProduct) Loyalty() ProductLoyalty {
 	}
 }
 
+func (avp ActiveVariantProduct) AvailableLoyalties() []ProductLoyalty {
```
Comment (tessig) [inline product/interfaces/graphql/product/dto/productdto_activevariantproduct.go:92]: Why don't we just call `AvailableLoyalties` on the `ActiveVariant()`?

Comment (IvanMaidurov) [reply product/interfaces/graphql/product/dto/productdto_activevariantproduct.go:92]: done in the same way as single loyalty done, and available loyalties still gonna need packing into ```ProductLoyalty```

File: product/domain/productBasics.go (hunk 1 above)
Comment (IvanMaidurov) [reply product/domain/productBasics.go]: done

File: product/domain/productBasics.go [hunk 3]
```diff
@@ -85,11 +85,12 @@ type (
...
-		SaleableFrom    time.Time
-		SaleableTo      time.Time
-		ActivePrice     PriceInfo
-		AvailablePrices []PriceInfo
+		IsSaleable         bool
+		SaleableFrom       time.Time
+		SaleableTo         time.Time
+		ActivePrice        PriceInfo
+		AvailablePrices    []PriceInfo
+		ActiveLoyaltyPrice LoyaltyPriceInfo
```
Comment (carstendietrich) [inline product/domain/productBasics.go:93]: What's the reason behind touching the domain model? thought we would only expose the already existing `LoyaltyPrices []LoyaltyPriceInfo`slice via graphql.

File: cart/domain/decorator/cartDecorator.go [hunk 4]
```diff
@@ -49,7 +49,8 @@ type (
 		Group          string
 	}
 
-	ctxKey struct{}
+	ctxKey         struct{}
```
Comment (tessig) [inline cart/domain/decorator/cartDecorator.go:52]: let's go for cartCtxKey then

File: product/domain/productBasics.go [hunk 5]
```diff
@@ -417,11 +418,10 @@ func (p Saleable) IsSaleableNow() bool {
 
 // GetLoyaltyPriceByType returns the loyaltyentry that matches the type
 func (p Saleable) GetLoyaltyPriceByType(ltype string) (*LoyaltyPriceInfo, bool) {
-	for _, lp := range p.LoyaltyPrices {
-		if lp.Type == ltype {
-			return &lp, true
-		}
+	if p.ActiveLoyaltyPrice.Type == ltype {
+		return &p.ActiveLoyaltyPrice, true
```
Comment (tessig) [inline product/domain/productBasics.go:422]: How is ActiveLoyaltyPrice supposed to behave when we have multiple loyalty programs?

File: cart/domain/decorator/cartDecorator.go (hunk 4 above)
Comment (IvanMaidurov) [reply cart/domain/decorator/cartDecorator.go]: done

**Review by carstendietrich — APPROVED** [review]
lgtm

File: product/domain/productBasics.go (hunk 3 above)
Comment (carstendietrich) [inline product/domain/productBasics.go:93]:
```suggestion
		ActiveLoyaltyPrice *LoyaltyPriceInfo
```

Shouldnt this be a pointer for products that don't have a loyalty price?

**Review by carstendietrich — CHANGES_REQUESTED** [review]

File: product/domain/productBasics.go [hunk 6]
```diff
@@ -443,13 +449,19 @@ func (p Saleable) generateLoyaltyChargeSplit(valuedPriceToPay *priceDomain.Price
 	requiredCharges := make(map[string]priceDomain.Charge)
 	remainingMainChargeValue := valuedPriceToPay.Amount()
 
+	if p.ActiveLoyaltyPrice == nil || !p.ActiveLoyaltyPrice.GetFinalPrice().IsPositive() {
+		return buildCharges(requiredCharges, *remainingMainChargeValue, *valuedPriceToPay)
+	}
+
 	// getLoyaltyCharge - private func that returns the loyaltyCharge of the given type. making sure the currentlyRemainingMainChargeValue is not exceeded
 	getValidLoyaltyCharge := func(loyaltyAmountWishedToSpent big.Float, loyaltyPrice LoyaltyPriceInfo, chargeType string, currentlyRemainingMainChargeValue *big.Float) priceDomain.Charge {
```
Comment (tessig) [inline product/domain/productBasics.go:457]: shall we move this inline func out to the package for better readability?

**Review by carstendietrich — APPROVED** [review]

**Pull request merged**